module/os/linux/zfs/zvol_os.c

```diff
@@ -1179,19 +1404,42 @@ int
 zvol_init(void)
 {
 	int error;
-	int threads = MIN(MAX(zvol_threads, 1), 1024);
+
+	if (zvol_threads == 0) {
+		zvol_actual_threads = num_online_cpus();
+	} else {
+		zvol_actual_threads = MIN(MAX(zvol_threads, 1), 1024);
+	}
 
 	error = register_blkdev(zvol_major, ZVOL_DRIVER);
 	if (error) {
 		printk(KERN_INFO "ZFS: register_blkdev() failed %d\n", error);
 		return (error);
 	}
-	zvol_taskq = taskq_create(ZVOL_DRIVER, threads, maxclsyspri,
-	    threads * 2, INT_MAX, TASKQ_PREPOPULATE | TASKQ_DYNAMIC);
+
+#ifdef	HAVE_BLK_MQ
+	if (zvol_blk_mq_queue_depth == 0) {
+		zvol_actual_blk_mq_queue_depth = BLKDEV_DEFAULT_RQ;
+	} else {
+		zvol_actual_blk_mq_queue_depth =
+		    MAX(zvol_blk_mq_queue_depth, BLKDEV_MIN_RQ);
+	}
+
+	/* Create a kmem cache for all our blk-mq work items */
+	blk_mq_cache = kmem_cache_create("blk_mq_cache",
+	    sizeof (zv_work_t), 0, NULL, NULL, NULL, NULL, NULL, 0);
+	if (!blk_mq_cache)
```

Comment:
You need to unregister the blkdev here on error.

```suggestion
	if (blk_mq_cache == NULL) {
		unregister_blkdev(zvol_major, ZVOL_DRIVER);
		return (-ENOMEM)
	}
```